Add support for user-space framebuffer(`/dev/fb0`)

Tangzh33 opened the pull request:
This PR introduces user-space framebuffer access through `/dev/fb0`, allowing applications to interact directly with the display framebuffer. With this implementation, key capabilities are enabled for user apps:

- Basic framebuffer operations (read/write pixels)
- Screen resolution and color depth information
- Memory-mapped access for high-performance graphics
- Simple display control and manipulation

Note: For simplicity, we currently only support the GOP (Graphics Output Protocol) Framebuffer. In other words, the GOP Framebuffer is initialized by the EFI, and the device capabilities are limited. Currently, it doesn't support some advanced `ioctls`, such as FBIO_BLANK. The implementation focuses on core functionality first, making it suitable for basic graphics applications. Future enhancements can add more advanced features as needed.

Lastly, this PR is built upon PR #2099 and provides vital support for Issue #2008. Here are some materials for quick reference: 

1. https://wiki.osdev.org/GOP
2. https://docs.kernel.org/fb/efifb.html
3. https://github.com/torvalds/linux/blob/master/drivers/video/fbdev/efifb.c

lrh2000 commented:
- What's the point of the commit [`Make GOP framebuffer region marked with Framebuffer type`](https://github.com/asterinas/asterinas/pull/2216/commits/7269b3c1606c80540d97003ec6c2020e1a5c8096)? It has already been marked as a framebuffer. It seems like you're doing random work to figure out how memory regions overlap, but you don't need to because the final `region.into_non_overlapping()` will automatically work things out.

 - What's the point of the commit [`Add a framebuffer allocator`](https://github.com/asterinas/asterinas/pull/2216/commits/c3883c02988b99d2786b708634e0c204f71f53d5)? The framebuffer region has already been reserved in the allocator after https://github.com/asterinas/asterinas/pull/2047.

I see no point of the above commits. I suspect they're ugly (sorry) workarounds for bugs that were fixed in the mainline a while ago. Please try to drop those commits and see if things still work.

---

Also, the commit [`Add color inforomation while initializing the GOP framebuffer`](https://github.com/asterinas/asterinas/pull/2216/commits/a61dcf8828edb94a33ac48ce0310c53d99e17948) isn't implemented in a correct way. You should rely on the `pixel_format: PixelFormat` field to retrieve the color information. Do not add new fields, they're pointless. (Of course, the current supported `PixelFormat` is limited, you can enhance it, but it's a totally different story and should not even belong to this PR).
```rust
// In `comps/framebuffer`,
impl Framebuffer {
    fn pixel_format(&self) -> PixelFormat;
}

// In `kernel`,
let pixel_format = fb.pixel_format();
let red_bitfield = todo!("fill this information according to `pixel_format`");
let blue_bitfield = todo!("fill this information according to `pixel_format`");
let green_bitfield = todo!("fill this information according to `pixel_format`");
```

Tangzh33 commented:
This pull request builds on PR #2099, which has already been merged. Therefore, this PR is ready for review:-)

However, upon reviewing the lessons learned from the iterations of PR #2099, I identified some recurring issues in the current PR. I am currently addressing these issues and refactoring the code structure, with the changes expected to be completed by the end of today. Please review the PR afterwards. Thank you!

lrh2000 reviewed (commented):
Thanks for the work!

I previously wondered how to manage framebuffer ownership, particularly how to disable framebuffer console functions when the userspace program (e.g., Xorg) uses `/dev/fb0`. After investigating, I found that the framebuffer console must be disabled manually by the userspace program using [`KDSETMODE(KD_GRAPHICS)`](https://man7.org/linux/man-pages/man2/ioctl_kd.2.html#KDSETMODE), which is part of virtual terminal (VT) support. That is not related to the PR at all.

Tangzh33 commented on ostd/src/boot/mod.rs:
```diff
@@ -73,6 +73,23 @@ pub struct BootloaderFramebufferArg {
     pub bpp: usize,
+    /// The color information of the framebuffer.
+    /// The size of the red channel.
+    pub red_size: u8,
```
Sure, I have removed this unrelated commit in the latest version.

Tangzh33 commented on kernel/src/fs/utils/ioctl.rs:
```diff
@@ -46,4 +46,46 @@ pub enum IoctlCmd {
+    ///
+    /// # Arguments
+    /// - Input: none
+    /// - Output: [`FbFixScreenInfo`] structure with hardware-specific information
```
These comments appear somewhat misplaced here. I have relocated them to the documents of their corresponding handlers and rewritten them in a more natural language style.

lrh2000 commented on kernel/src/device/fb.rs:
```diff
@@ -246,9 +381,47 @@ impl FileIo for FbHandle {
+            IoctlCmd::GETVSCREENINFO => self.handle_get_var_screen_info(arg),
+            IoctlCmd::GETFSCREENINFO => self.handle_get_fix_screen_info(arg),
+            IoctlCmd::PUTVSCREENINFO => {
+                // The framebuffer we are working with is initialized by UEFI
+                //  services, which do not support changing **ANY** settings.
+                //  Therefore, we simply copy the current settings back to
+                //  userspace without making any changes. This behavior is
+                //  consistent with Linux's efifb driver.
```
```rust
        // EFI framebuffers do not support changing settings. Linux will return
        // the old settings to user space and succeed.
        // Reference: <https://elixir.bootlin.com/linux/v6.17/source/drivers/video/fbdev/core/fbmem.c#L276-L279>.
```

lrh2000 commented on kernel/src/device/fb.rs:
```diff
@@ -15,6 +19,15 @@ use crate::{
+const DEFAULT_RIGHT_MARGIN: u32 = 32;
+const DEFAULT_UPPER_MARGIN: u32 = 16;
+const DEFAULT_LOWER_MARGIN: u32 = 4;
+const DEFAULT_VSYNC_LEN: u32 = 4;
```
Move the definitions of these constants to `handle_get_var_screen_info`.

lrh2000 commented on kernel/src/device/fb.rs:
```diff
@@ -311,6 +272,7 @@ impl FbHandle {
     }
 
     /// Handles the [`IoctlCmd::GETFSCREENINFO`] ioctl command.
+    /// Takes no input and outputs a [`FbFixScreenInfo`] structure with hardware-specific information.
```
Move this to the correct commit.

I think the previous format is okay, perhaps it's actually clearer than the current format:
```rust
    /// Handles the [`IoctlCmd::GETFSCREENINFO`] ioctl command.
    ///
    /// Arguments:
    ///  - Input: [`FbFixScreenInfo`].
    ///  - Output: None.
```

lrh2000 commented on kernel/src/device/fb.rs:
```diff
@@ -325,6 +287,109 @@ impl FbHandle {
+        let start = cmap_user.start as usize;
+        let len = cmap_user.len as usize;
+
+        if start >= cmap.red.len() || start + len > cmap.red.len() {
```
```suggestion
        if start >= cmap.red.len() || len > cmap.red.len() - start {
```

`start + len` can overflow, which will make the kernel panic in debug builds.

lrh2000 commented on kernel/src/device/fb.rs:
```diff
@@ -325,6 +287,109 @@ impl FbHandle {
+        let mut cmap = self.framebuffer.cmap().lock();
+
+        // Ensure the colormap has enough space
+        let required_len = start + len;
```
This can overflow. This need to be checked against `256` earlier (similar to my previous comment).

lrh2000 commented on kernel/src/device/fb.rs:
```diff
@@ -325,6 +287,109 @@ impl FbHandle {
+        }
+
+        // Size check to prevent excessive memory allocation
+        if cmap_user.len > 256 {
```
This should be a constant from the `aster-framebuffer`.

lrh2000 commented on kernel/src/device/fb.rs:
```diff
@@ -325,6 +287,109 @@ impl FbHandle {
+        let len = cmap_user.len as usize;
+
+        if start >= cmap.red.len() || start + len > cmap.red.len() {
+            return_errno_with_message!(Errno::EINVAL, "colormap index out of bounds");
```
```suggestion
            return_errno_with_message!(Errno::EINVAL, "the color map index is out of bounds");
```

lrh2000 commented on kernel/src/device/fb.rs:
```diff
@@ -325,6 +287,109 @@ impl FbHandle {
+            return Ok(0);
+        }
+
+        // Size check to prevent excessive memory allocation
+        if cmap_user.len > 256 {
+            return_errno_with_message!(Errno::EINVAL, "colormap too large");
```
```suggestion
        // Check the size to prevent excessive memory allocation.
        if cmap_user.len > 256 {
           return_errno_with_message!(Errno::EINVAL, "the color map size is too large");
```

lrh2000 commented on kernel/src/device/fb.rs:
```diff
@@ -72,20 +73,52 @@ impl FileIo for Fb {
+        let mut offset = self.offset.lock();
+
+        if *offset >= buffer_size {
+            return_errno_with_message!(Errno::ENOSPC, "write beyond framebuffer size");
```
```suggestion
            return_errno_with_message!(Errno::ENOSPC, "the write offset is beyoud the framebuffer size");
```

Are you sure this shouldn't be an `OK (0)`?

Tangzh33 replied:
I believe that when writing to the framebuffer beyond its capacity, the return value should be `ENOSPC`. I have tested this on Linux and included it in the CI test :-)

lrh2000 commented on kernel/comps/framebuffer/src/framebuffer.rs:
```diff
@@ -132,6 +192,120 @@ impl FrameBuffer {
+        cmap_len: u32,
+    ) -> Result<i32> {
+        if regno >= cmap_len {
+            return Ok(1); // Return 1 for invalid regno, matching Linux behavior
```
You ignore the error code here:
```rust
       let _ = self.framebuffer.set_color_regs(
```

So, there is nothing that matches the Linux behavior. As I've said before, it isn't reasonable to match the Linux code at the implementation level, and that wasn't our goal.

lrh2000 commented on kernel/src/device/fb.rs:
```diff
@@ -72,20 +73,52 @@ impl FileIo for Fb {
+        let mut offset = self.offset.lock();
+
+        if *offset >= buffer_size {
+            return_errno_with_message!(Errno::ENOSPC, "write beyond framebuffer size");
```
As we discussed, testing the correctness of the framebuffer is difficult. However, while reviewing the latest version of this PR, I realized there are some things we can test.
 - For instance, we can test the color map and certain error codes, such as the one I mentioned above (`ENOSPC`).
 - On the other hand, the previous PR adds support for mapping MMIO memory to userspace. However, it has not been thoroughly tested by regression tests.
   - For example, we can map the framebuffer, write the framebuffer, then do some partial `mremap` or do a `fork`, and finally check the contents.
 
I'm not sure if you're familar with the test framework, i.e., [`test/src/apps/test.h`](test/src/apps/test.h), but there are lots of users under `test/src/apps` (you can try to find them by [searching `#include "../test.h"`](https://github.com/search?q=repo%3Aasterinas%2Fasterinas+%22%23include+%5C%22..%2Ftest.h%5C%22%22&type=code)). Is it possible to add one such test, add the program name to [`test/src/apps/scripts/*.sh`](https://github.com/asterinas/asterinas/tree/9e4a67879f601c6b25f9394e9919b88ef34c7cfd/test/src/apps/scripts), and see if it can run in CI?

lrh2000 reviewed (commented):
Thanks for the revision. There are some design problems about the color map implementation. The current implementation is really odd to me.

Tangzh33 commented:
Thank you for your comments, and sorry for the delay. I have been focused on booting XFCE on the laptop, and fortunately, it runs smoothly as expected. I appreciate your help in our numerous private discussions.

I have refactored the unreasonable interface between `aster-framebuffer` and `aster-nix`. Additionally, I added a CI test for the EFI framebuffer's `mmap` and `ioctl` interfaces. To support the functionality of `lseek` (used in the CI test to evaluate write-beyond-size behavior), as well as `read` and `write` with an offset, I added more fields to the `FileIO` trait and updated the related functions accordingly. This PR is ready for review again, please drop your comments at your covenience.

lrh2000 commented on kernel/src/device/fb.rs:
```diff
@@ -1,6 +1,8 @@
 // SPDX-License-Identifier: MPL-2.0
 
-use aster_framebuffer::{FrameBuffer, PixelFormat, FRAMEBUFFER};
+use core::mem::size_of;
```
I do not think https://github.com/asterinas/asterinas/pull/2216#discussion_r2389810188 is resolved.

lrh2000 commented on kernel/src/device/fb.rs:
```diff
@@ -72,20 +73,52 @@ impl FileIo for Fb {
+
+        // Read from the framebuffer at the current offset.
+        // Limit the writer to avoid over-reading when the user buffer is
+        // larger than the remaining framebuffer size.
```
I do not think https://github.com/asterinas/asterinas/pull/2216#discussion_r2389838969 is resolved.

The current implementation of `read`/`read_at`/`write`/`write_at` is unnecessarily complex. I cannot understand it at all.

Does the proposed version in https://github.com/asterinas/asterinas/pull/2216#discussion_r2389838969 work?

lrh2000 commented on kernel/src/fs/inode_handle/mod.rs:
```diff
@@ -114,7 +114,11 @@ impl InodeHandle_ {
-        do_seek_util(self.path.inode(), &self.offset, pos)
+        let new_offset = do_seek_util(self.path.inode(), &self.offset, pos)?;
+        if let Some(ref file_io) = self.file_io {
+            file_io.set_offset(new_offset);
```
In race conditions, this can cause inconsistent state between the `InodeHandle` and the `FileIo` object.

After some thought, I think perhaps we should make `FileIo` a super-trait of `Inode` and refactor the `InodeHandle` implementation instead. I will try later to see if my idea works.

Tangzh33 replied:
Oh, I missed the "EDIT" part; I will check it later.

lrh2000 commented on kernel/src/device/fb.rs:
```diff
@@ -28,6 +28,177 @@ pub struct FbHandle {
+}
+
+/// Converts pixel format to framebuffer bitfields for Linux compatibility.
+#[rustfmt::skip]
```
This should be on `match`. Or is there a problem preventing it from working?
```rust
    #[rustfmt::skip]
    match pixel_format {
    }
```

Tangzh33 replied:
It seems that the compiler would not be happy with the ` #[rustfmt::skip]` on expressions. Moving the attributes to the function would simply solve this problem.

``` bash
error[E0658]: attributes on expressions are experimental
  --> kernel/src/device/fb.rs:46:9
   |
46 |         #[rustfmt::skip]
   |         ^^^^^^^^^^^^^^^^
   |
   = note: see issue #15701 <https://github.com/rust-lang/rust/issues/15701> for more information
   = help: add `#![feature(stmt_expr_attributes)]` to the crate attributes to enable
   = note: this compiler was built on 2025-01-31; consider upgrading it if it is out of date

For more information about this error, try `rustc --explain E0658`.
error: could not compile `aster-nix` (lib) due to 1 previous error
[Error]: Command RUSTFLAGS=" --check-cfg cfg(ktest) --cfg ktest" "cargo" "clippy" "-Zbuild-std=core,alloc,compiler_builtins" "-Zbuild-std-features=compiler-builtins-mem" "--target" "x86_64-unknown-none" "--" "-D" "warnings" failed with status: ExitStatus(unix_wait_status(25856))
make: *** [Makefile:360: check] Error 1
```

lrh2000 commented on test/src/apps/framebuffer/framebuffer_tests.c:
```diff
@@ -0,0 +1,348 @@
+		}
+
+		// Ensure data is flushed
+		if (msync(mapped, map_len, MS_SYNC) != 0) {
```
Is the code generated by LLM? Please remove unnecessary code... Why we need `msync` here?

lrh2000 commented on kernel/comps/framebuffer/src/framebuffer.rs:
```diff
@@ -129,9 +180,56 @@ impl FrameBuffer {
+    }
+
+    /// Returns the current length of the color map.
+    pub fn color_map_len(&self) -> usize {
```
This method is not used? Remove it if it has no users.

lrh2000 reviewed (commented):
Thanks for the revision!

Most of the Rust code is fine, but I do not think the current approach to extending `FileIo` to support offsets is correct (it's racy after all). I'll try some other approaches later to see if they work (see https://github.com/asterinas/asterinas/pull/2216#discussion_r2493399765).

Many comments are left on the regression test because I suspect that it was generated by an LLM, which makes it overly verbose and odd-looking. At least, the style is very different from existing regression tests.

Tangzh33 commented:
@lrh2000 Thank you for your reviews!

Based on our discussion, this PR should now be built upon PR #2564 to properly address the issues in the [comments](https://github.com/asterinas/asterinas/pull/2216#discussion_r2493399765). In this iteration, I will leave the problematic parts untouched and focus on fixing the remaining issues. This PR is now ready for review again. Please feel free to share your comments whenever convenient.

lrh2000 commented on test/src/apps/devfs/framebuffer.c:
```diff
@@ -0,0 +1,214 @@
+{
+	if (fb_fd >= 0) {
+		close(fb_fd);
+		fb_fd = -1;
```
This is a good habit but I don't think it's necessary to do this in test files. To align with the existing code, I'd prefer:
````rust
    CHECK(close(fb_fd));
```

lrh2000 commented on test/src/apps/devfs/framebuffer.c:
```diff
@@ -0,0 +1,214 @@
+
+	pid_t pid = CHECK(fork());
+	if (pid == 0) {
+		// Child process: avoid TEST_* so the parent sees the status via waitpid
```
Yes, so please use `CHECK`.
```suggestion
CHECK_WITH(memcmp(mapped, pattern, map_len), _ret == 0);
```

lrh2000 commented on test/src/apps/devfs/framebuffer.c:
```diff
@@ -0,0 +1,214 @@
+	memcpy(mapped, pattern, map_len);
+	TEST_RES(memcmp(mapped, pattern, map_len), _ret == 0);
+
+	uint8_t *remapped = TEST_RES((uint8_t *)mremap(mapped, map_len, map_len,
```
Use `TEST_SUCC` for `mremap` is okay because `TEST_SUCC` will check the error code to determine if the system call succeeds.

lrh2000 commented on test/src/apps/devfs/framebuffer.c:
```diff
@@ -0,0 +1,214 @@
+	}
+
+	uint8_t *mapped =
+		TEST_RES((uint8_t *)mmap(NULL, map_len, PROT_READ | PROT_WRITE,
```
Same here.

lrh2000 commented on kernel/src/device/fb.rs:
```diff
@@ -0,0 +1,94 @@
+    }
+
+    fn open(&self) -> Option<Result<Arc<dyn FileIo>>> {
+        let framebuffer = FRAMEBUFFER.get()?.clone();
```
This is incorrect. See my patch below. But anyway, the signature of this method will be changed in https://github.com/asterinas/asterinas/pull/2564, so this problem won't happen after that.

Tangzh33 replied:
I have fixed it in the current version and will handle this carefully in future rebases.

From my perspective, here is the reason why the CI pass even with a buggy open: currently, the device file (`/dev/fb0`) of the EFI framebuffer is hardcoded with device probing. If the boot service does not supply a framebuffer, the device file will not be created. The buggy device-specific `open` will not be triggered when the framebuffer is absent because the syscall will return the error number `ENOENT`. Therefore, without the `/dev/fb0`, the CI will fail silently and still pass.

lrh2000 commented on kernel/src/device/fb.rs:
```diff
@@ -74,9 +74,55 @@ impl FileIo for Fb {
+        let buffer_size = self.framebuffer.io_mem().size();
+        let mut offset = self.offset.lock();
+
+        if *offset >= buffer_size {
```
Prefer to be consistent. Check this using `write.avail()` instead, just like how we implement `read()`:
```patch
diff --git a/kernel/src/device/fb.rs b/kernel/src/device/fb.rs
index 21b2a96f9..6c13cf7aa 100644
--- a/kernel/src/device/fb.rs
+++ b/kernel/src/device/fb.rs
@@ -468,12 +473,14 @@ impl FileIo for FbHandle {
             return Ok(0);
         }
 
-        let buffer_size = self.framebuffer.io_mem().size();
-        if offset >= buffer_size {
-            return_errno_with_message!(Errno::ENOSPC, "write beyond framebuffer size");
-        }
-
         let mut writer = self.framebuffer.io_mem().writer();
+
+        if offset >= writer.avail() {
+            return_errno_with_message!(
+                Errno::ENOSPC,
+                "the write offset is beyond the framebuffer size"
+            );
+        }
         writer.skip(offset);
 
         let mut writer = writer.to_fallible();
@@ -491,14 +498,15 @@ impl FileIo for FbHandle {
             return Ok(0);
         }
 
-        let buffer_size = self.framebuffer.io_mem().size();
+        let mut writer = self.framebuffer.io_mem().writer();
         let mut offset = self.offset.lock();
 
-        if *offset >= buffer_size {
-            return_errno_with_message!(Errno::ENOSPC, "write beyond framebuffer size");
+        if *offset >= writer.avail() {
+            return_errno_with_message!(
+                Errno::ENOSPC,
+                "the write offset is beyond the framebuffer size"
+            );
         }
-
-        let mut writer = self.framebuffer.io_mem().writer();
         writer.skip(*offset);
 
         let mut writer = writer.to_fallible();
@@ -507,8 +515,8 @@ impl FileIo for FbHandle {
             Err((err, 0)) => return Err(err.into()),
             Err((_err, len)) => len,
         };
-
         *offset += len;
+
         Ok(len)
     }
 ```

lrh2000 commented on kernel/src/device/fb.rs:
```diff
@@ -245,6 +245,30 @@ impl FileIo for Fb {
+
+    /// Writes an array of `u16` color map values to userspace.
+    pub fn write_color_maps_to_user(addr: usize, data: &[u16]) -> Result<()> {
+        if addr == 0 {
```
In fact, we should only check this for the transparency channel. Read the Linux implementation first before making a decision: https://elixir.bootlin.com/linux/v6.17.7/source/drivers/video/fbdev/core/fbcmap.c#L281-L284

The same applies for `read_color_maps_from_user`.

lrh2000 reviewed (commented):
Thanks for addressing the previous comments!

The new comments should all be minor.

lrh2000 commented on kernel/src/device/fb.rs:
```diff
@@ -0,0 +1,86 @@
+pub struct Fb;
+
+#[derive(Debug)]
+pub struct FbHandle {
```
I don't think `Fb` and `FbHandle` needs to be `pub`?

lrh2000 reviewed (commented):
Thanks for the updates. These should be my final comments, and they should be very minor.

lrh2000 commented on test/src/apps/devfs/framebuffer.c:
```diff
@@ -0,0 +1,193 @@
+#define _GNU_SOURCE
+
+// SPDX-License-Identifier: MPL-2.0
```
I am not aware of a case in which we placed the license line after the `#define _GNU_SOURCE` line.

https://github.com/asterinas/asterinas/blob/2034055f90607f67f04de044ee88dd52a488d612/test/src/apps/clone3/clone_process.c#L1-L4

Prefer to follow the style of other tests, see: https://github.com/search?q=repo%3Aasterinas%2Fasterinas+_GNU_SOURCE&type=code&p=1

tatetian commented on kernel/src/device/fb.rs:
```diff
@@ -0,0 +1,508 @@
+        let Some(framebuffer) = FRAMEBUFFER.get() else {
+            return Err(Error::with_message(
+                Errno::ENODEV,
+                "the framebuffer device is not present",
```
Thanks to the new char device registration mechanism, device inodes for char devices are created automatically for every registered char devices. So I really wanted to mean "do not register the framebuffer device in the first place".

lrh2000 replied:
But will `mknod` succeed if we don't register the framebuffer device?

https://github.com/asterinas/asterinas/blob/808d298938ef53dab6057fa54c8de96360e4ea46/kernel/src/device/mod.rs#L85-L88

I thought this should not fail, so we should always register the device regardless whether the device exists.

---

However, upon further consideration, it may be better to treat this as a separate issue. According to Linux behavior, regardless of whether the device numbers make sense, `mknod` should always succeed.

I performed an experiment under QEMU. Linux does not seem to create `fb0` if there is no framebuffer.

tatetian replied:
@Tangzh33 To be more concrete, instead of checking if `FRAMEBUFFER` in the open operation, we should do the check when registering the FB char device.

```rust
pub(super) fn init_in_first_kthread() {
    if FRAMEBUFFER.get().is_none() else {
        return;
    }

    char::register(Arc::new(Fb)).expect("failed to register framebuffer char device");
}
```

Tangzh33 replied:
Sure, I am currently working on this solution, and the behavior will be updated in the next iteration.

lrh2000 commented on kernel/src/device/fb.rs:
```diff
@@ -30,6 +30,177 @@ struct FbHandle {
+    pub reserved: [u16; 2],
+}
+
+/// Framebuffer colormap structure for userspace communication.
```
```suggestion
/// Framebuffer colormap structure for userspace communication; `struct fb_cmap` in Linux.
```

lrh2000 reviewed (approved):
Two small nits. Otherwise LGTM, thanks for addressing all the comments.
